private/bufpkg/bufimage/bufimageutil/image_filter.go

```diff
@@ -0,0 +1,672 @@
+	method *descriptorpb.MethodDescriptorProto,
+) (*descriptorpb.MethodDescriptorProto, bool, error) {
+	if !b.closure.hasType(method, b.options) {
+		return nil, true, nil
+	}
+	newOptions, changed, err := remapMessage(nil, append(sourcePath, methodOptionsTag), method.GetOptions(), b.remapOptions)
+	if err != nil {
+		return nil, false, err
+	}
+	if !changed {
+		return method, false, nil
+	}
+	newMethod := maybeClone(method, b.options)
+	newMethod.Options = newOptions
+	return newMethod, true, nil
+}
+
+func (b *sourcePathsBuilder) remapField(
+	sourcePathsRemap *sourcePathsRemapTrie,
+	sourcePath protoreflect.SourcePath,
+	field *descriptorpb.FieldDescriptorProto,
+) (*descriptorpb.FieldDescriptorProto, bool, error) {
+	if field.Extendee != nil {
+		// Extensions are filtered by type.
+		if !b.closure.hasType(field, b.options) {
+			return nil, true, nil
+		}
+	}
+	switch field.GetType() {
+	case descriptorpb.FieldDescriptorProto_TYPE_ENUM,
+		descriptorpb.FieldDescriptorProto_TYPE_MESSAGE,
+		descriptorpb.FieldDescriptorProto_TYPE_GROUP:
+		typeName := protoreflect.FullName(strings.TrimPrefix(field.GetTypeName(), "."))
+		typeInfo := b.imageIndex.ByName[typeName]
+		if !b.closure.hasType(typeInfo.element, b.options) {
+			return nil, true, nil
+		}
+	case descriptorpb.FieldDescriptorProto_TYPE_DOUBLE,
+		descriptorpb.FieldDescriptorProto_TYPE_FLOAT,
+		descriptorpb.FieldDescriptorProto_TYPE_INT64,
+		descriptorpb.FieldDescriptorProto_TYPE_UINT64,
+		descriptorpb.FieldDescriptorProto_TYPE_INT32,
+		descriptorpb.FieldDescriptorProto_TYPE_FIXED64,
+		descriptorpb.FieldDescriptorProto_TYPE_FIXED32,
+		descriptorpb.FieldDescriptorProto_TYPE_BOOL,
+		descriptorpb.FieldDescriptorProto_TYPE_STRING,
+		descriptorpb.FieldDescriptorProto_TYPE_BYTES,
+		descriptorpb.FieldDescriptorProto_TYPE_UINT32,
+		descriptorpb.FieldDescriptorProto_TYPE_SFIXED32,
+		descriptorpb.FieldDescriptorProto_TYPE_SFIXED64,
+		descriptorpb.FieldDescriptorProto_TYPE_SINT32,
+		descriptorpb.FieldDescriptorProto_TYPE_SINT64:
+	default:
+		return nil, false, fmt.Errorf("unknown field type %d", field.GetType())
+	}
+	newOption, changed, err := remapMessage(nil, append(sourcePath, fieldOptionsTag), field.GetOptions(), b.remapOptions)
+	if err != nil {
+		return nil, false, err
+	}
+	if !changed {
+		return field, false, nil
+	}
+	newField := maybeClone(field, b.options)
+	newField.Options = newOption
+	return newField, true, nil
+}
+
+func (b *sourcePathsBuilder) remapOptions(
+	sourcePathsRemap *sourcePathsRemapTrie,
+	optionsPath protoreflect.SourcePath,
+	optionsMessage proto.Message,
+) (proto.Message, bool, error) {
+	if optionsMessage == nil {
+		return nil, false, nil
+	}
+	var newOptions protoreflect.Message
+	options := optionsMessage.ProtoReflect()
+	numFieldsToKeep := 0
+	options.Range(func(fd protoreflect.FieldDescriptor, val protoreflect.Value) bool {
+		if !b.closure.hasOption(fd, b.imageIndex, b.options) {
+			// Remove this option.
```

Comment:
Why bother? In previous formulation, we didn't bother trying to strip away options that were not included in the filtered image (which could happen if `includeCustomOptions` was set to false). They can be left as unrecognized bytes. I think the same is true here: even for explicitly excluded option types, do we really need to purge their values?

Reply:
Is it a performance thing to not remove them? This was added for the explicit options removing, but that's no longer needed. I like that it's cleaned up and doesn't need to re-serialize the message to convert them to unrecognized.

Reply:
In that last commit, I think it's safe to simply remove `remapOptions` entirely.

Reply:
Fixed in [b4c1f56](https://github.com/bufbuild/buf/pull/3624/commits/b4c1f56931d6cae7d0661fc6d3e53afbf8fb5e28). Updated the import logic to check for changes in [c6b0e6a](https://github.com/bufbuild/buf/pull/3624/commits/c6b0e6abc90df125875ca9845f0c822072fbf583).

The file descriptor produced failed to pass checks from the previous filter:
```
inclusion of type "a.v1.Foo": cannot find ext no 50002 on google.protobuf.MessageOptions
```
This option check has been relaxed in [1c78fc9](https://github.com/bufbuild/buf/pull/3624/commits/1c78fc96796f5c32e835cebef866678717371b3e). This allows filters to be applied in consecutively.